basesandbox/src/ipc/domain_socket.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use std::cell::RefCell;
+use std::os::unix::net::UnixDatagram;
+use std::path::Path;
+use std::sync::Arc;
+
+const TEMPORARY_PATH: &str = "/tmp";
+
+pub struct DomainSocketLinker {
+    address_server: String,
+    address_client: String,
+}
+
+impl TwoWayInitialize for DomainSocketLinker {
+    type Server = DomainSocket;
+    type Client = DomainSocket;
+
+    fn new(_name: String) -> Self {
+        let address_server = format!("{}/{}", TEMPORARY_PATH, generate_random_name());
+        let address_client = format!("{}/{}", TEMPORARY_PATH, generate_random_name());
+        DomainSocketLinker {
+            address_server,
+            address_client,
+        }
+    }
+
+    fn create(&self) -> (Vec<u8>, Vec<u8>) {
+        (
+            serde_cbor::to_vec(&(&self.address_server, &self.address_client)).unwrap(),
+            serde_cbor::to_vec(&(&self.address_client, &self.address_server)).unwrap(),
+        )
+    }
+}
+
+impl Drop for DomainSocketLinker {
+    fn drop(&mut self) {
+        std::fs::remove_file(&self.address_server).unwrap();
+        std::fs::remove_file(&self.address_client).unwrap();
+    }
+}
+
+// TODO: Use stream instead of datagram.
+struct SocketInternal(UnixDatagram);
+impl Drop for SocketInternal {
+    fn drop(&mut self) {
+        self.0.shutdown(std::net::Shutdown::Both).unwrap();
+    }
+}
+
+pub struct DomainSocketSend {
+    address_dst: String,
+    socket: Arc<SocketInternal>,
+}
+
+impl IpcSend for DomainSocketSend {
+    fn send(&self, data: &[u8]) {
+        assert_eq!(self.socket.0.send_to(data, &self.address_dst).unwrap(), data.len());
+    }
+}
+
+pub struct DomainSocketRecv {
+    address_dst: String,
+    socket: Arc<SocketInternal>,
+    buffer: RefCell<Vec<u8>>,
+}
+
+impl IpcRecv for DomainSocketRecv {
+    type MyTerminate = Terminator;
```

Comment:
`My` prefix should be avoided outside examples, IMHO. It doesn't convey meaningful information. Call it `Terminate` or `TerminateRecv`.

Reply:
Oh sure. My mistake.